include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -863,6 +863,173 @@ __bypass_sycl_kernel_not_supported(const sycl::exception& __e)
         throw;
 }
 
+struct __scalar_load_op
+{
+    oneapi::dpl::__internal::__pstl_assign __assigner;
+    template <typename _IdxType1, typename _IdxType2, typename _SourceAcc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _SourceAcc __source_acc, _DestAcc __dest_acc) const
+    {
+        __assigner(__source_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_load
+{
+    static_assert(__vec_size <= 4, "Only vector sizes of 4 or less are supported");
+    std::size_t __full_range_size;
+    template <typename _IdxType, typename _LoadOp, typename... _Rngs>
+    void
+    operator()(/*__is_full*/ std::true_type, _IdxType __start_idx, _LoadOp __load_op, _Rngs&&... __rngs) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+
+    template <typename _IdxType, typename _LoadOp, typename... _Rngs>
+    void
+    operator()(/*__is_full*/ std::false_type, _IdxType __start_idx, _LoadOp __load_op, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __start_idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+};
+
+template <typename _TransformOp>
+struct __scalar_store_transform_op
```

Comment:
Another 5 cent of criticism)
What does the functor? it looks it does nothing.
As I can see it just re-calls  `_TransformOp` instance with 2 or 3 values.
Why whether dont use this `_TransformOp` directly, w/o wrapping it into `struct __scalar_store_transform_op`?

Reply:
Store Op in the `__vector_store` operation is called as `__store_op(__i, __start_idx + __i, __rngs...)` so we a need functor that matches that signature and handles indexing after unpacking the ranges.

`__transform` here will end up being something like `__transform_functor` which is a general brick used in scalar and vectorizable paths.

I guess it could be removed if `__vector_store` did not take a range pack and only had overloads for the currently used cases (unary, binary). I'm not sure which is better.